src/rgw/rgw_aws.cc

```diff
@@ -0,0 +1,259 @@
+#include "rgw_aws.h"
+#include <string>
+#include <iostream>
+#include <utility>
+#include "common/dout.h"
+#include <aws/core/Aws.h>
+#include <aws/sns/SNSClient.h>
+#include <aws/lambda/LambdaClient.h>
+#include <aws/lambda/model/InvokeRequest.h>
+#include <boost/lockfree/queue.hpp>
+#include <aws/core/utils/ARN.h>
+#include <aws/core/auth/AWSCredentials.h>
+#include <aws/sns/model/PublishRequest.h>
+
+#define dout_subsys ceph_subsys_rgw
+
+namespace rgw::aws {
+  static const int STATUS_OK = 1;
+  static const int STATUS_FULL = 2;
+  static const int STATUS_MANAGER_STOPPED = 3;
+  static const int STATUS_CLIENT_CLOSED = 4;
+
+  struct client_id_t {
+    const std::string accessKey;
+    const std::string accessSecret;
+    const std::string arn;
+    const bool useSSL;
+
+    client_id_t(const std::string accessKey, const std::string accessSecret,
+                const std::string arn, bool useSsl) : accessKey(accessKey),
+                                                      accessSecret(accessSecret),
+                                                      arn(arn), useSSL(useSsl) {}
+
+    bool operator==(const client_id_t &other) const {
+      return accessKey == other.accessKey &&
+             accessSecret == other.accessSecret &&
+             arn == other.arn && useSSL == other.useSSL;
+    }
+
+    struct hasher {
+      std::size_t operator()(const client_id_t &k) const {
+        return ((std::hash<std::string>()(k.accessKey)
+                 ^ (std::hash<std::string>()(k.accessSecret) << 1)) >> 1)
+               ^ (std::hash<std::string>()(k.arn) << 1)
+               ^ (std::hash<bool>()(k.useSSL) << 1);
+      }
+    };
+
+
+  };
+
+  struct message_t {
+    const AwsClient client;
+    const std::string resourceARN;
+    const std::string payload;
+
+    message_t(const AwsClient client, const std::string& resourceARN, const std::string& payload) :
+            client(client),
+            resourceARN(resourceARN),
+            payload(payload) {};
+  };
+
+  typedef std::unordered_map<client_id_t, AwsClient, client_id_t::hasher> ClientList;
+
+
+  typedef boost::lockfree::queue<message_t *, boost::lockfree::fixed_sized<true>> MessageQueue;
+
+  class Manager {
+  private:
+    MessageQueue messageQueue;
+    ClientList clientList;
+    std::thread runner;
+    bool stopped;
+    mutable std::mutex clientsLock;
+    CephContext *cct;
+    const ceph::coarse_real_clock::duration idle_time;
+
+    void publish_internal(message_t *msg) {
+      const std::unique_ptr<message_t> msg_owner(msg);
+
+      std::visit([&msg, this](auto &&arg) { this->publish_util(arg, msg->resourceARN, msg->payload); }, msg->client);
+    }
+
+    void publish_util(Aws::SNS::SNSClient *client, const string &resourceArn, const string &payload) {
+      ldout(cct, 1) << "Publishing.." << dendl;
```

Comment:
in `ldout` the second parameter is the priority.
0 - fatal
1 - error
2-19 warning/info
20 - debug
please use `1` for errors